# Week 3: Python Data Types and Variables

### Python Math
Let's start with some basics: run the following cells to see what they output

Note: a full list of operators supported in python can be found [here](https://www.tutorialspoint.com/python/python_basic_operators.htm)

In [1]:
2 + 1

3

In [2]:
36 - (4 + 1)**2

11

The outputs so far are all **integers** because they are round, whole numbers. 

What happens when run the previous cell with a decimal?

In [3]:
36 - (4.0 + 1)**2

11.0

We get the same answer numerically, but instead of an integer the resulting data type is called a **float**. 

This is the Pythonic name for a decimal value.

### Creating a Variable in Python


In [4]:
a = 1

In [5]:
print(a)

1


In [6]:
a + 1

2

Note that `a = 1` still.

In [7]:
print(a)

1


What happens to `a` when we assign `a = a + 2`?

In [8]:
a = a + 2
print(a)

3


Now we have overwritten the previous value of a with a new value!

#### Beware: Python is case sensitive

In [9]:
N = 20
n = 10
print(N, n)

20 10


#### Best practice: 
- Make your variable names descriptive! This will be very important later as our scripts get more complex
- Python convention is that variables are lower case and long variable names are separated by underscore (for example: `my_variable_name`)

### Data Types in Python

- **Numeric types:** integer, float
- **Text:** string
- **Collection of values:** list, array, tuple, dictionary
- **Logical:** boolean (pronounced boo-lee-uhn)

In [ ]:
list_of_ones = [int("1"), 1.0,  1, "1.0", ("first", 1), {"one": 1}, bool(1)]

Try to guess the output types before running the cell below:

In [11]:
for value in list_of_ones:
    print(type(value))

<class 'int'>
<class 'float'>
<class 'int'>
<class 'str'>
<class 'tuple'>
<class 'dict'>
<class 'bool'>


## Part 2: Indexing Variables

Lists are defined with square brackets and delineated by commas. 

They are distinct from another data type called **tuples** which are denoted by ( ) and are immutable (meaning they cannot be changed) once created. 
 
Let's learn how to index lists (and variables in general):

In [12]:
squares_list = [1, 4, 9, 16, 25, 36, 49, 64, 81, 100]

To access the list, we'll put brackets after the variable name. Let's try to get the first element in the list:

In [13]:
squares_list[1]

4

This is probably not what you were expecting. This is because in Python, the index starts at 0. So if we try this:

In [14]:
squares_list[0]

1

We'll get the "0th" element of the list. Now that we know this, we can get any element we like. Try to guess what this will output:

In [15]:
squares_list[4]

25

If we put in a number that's too large, Python will throw this error message:

In [16]:
squares_list[10]

IndexError: list index out of range

Now that we know how to access any element, let's try some more advanced stuff. To count from the back of the list, you can put negative numbers in the brackets:

In [ ]:
print(squares_list[-1])
print(squares_list[-3])

We can also access slices of the list. If we type a number, a colon, and then a second number, Python will make a list from between the two numbers:

In [ ]:
squares_list[1:4]
#Note that this includes index one but doesn't include index four

Lastly, we can change the 'step' of the slice. If we use a second colon, the number after it will give the 'step length':

In [ ]:
squares_list[1:10:2]

You can even use indexing to reverse a list!

In [ ]:
squares_list[::-1]

Great! Now we know how to index a list.

While they share similarities in indexing syntax, there are two key differences between strings and lists that we'll discuss in this module: 

1. Strings can only have characters––letters, numbers, and punctuation––whereas lists can have any data type
2. Python provides built-in **functions** that let you modify lists and strings.

For now, we'll start by discussing strings and tuples.

**Note**: strings and tuples are indexed exactly the same as lists.

In [ ]:
my_string = "I love Python!"
my_tuple = ("I", " ", "love", " ", "Python", "!")

Tuples are immutable, meaning they cannot be modified once created.

In [ ]:
# The following code produces an error
my_tuple[0] = "me"

Another thing to note: the following lines of code produce different variable types

In [ ]:
test_int = (1)
test_tuple = (1, )
test_list = [1]
for test in [test_int, test_tuple, test_list]:
    print(type(test))

Back to indexing! 

In [ ]:
# Index the elements at indices 0-5
print(my_string[0:6])

# Store the value at index 4 as a variable called python
python_string = my_tuple[4]
print(python_string)

# The following two lines of code are equivalent
print(python_string[0]) 
print(my_tuple[4][0])

In the first example, we sliced the 0th through the 5th index.

In the second example, we assigned a variable called `python_string` to the 5th element of the tuple. When we print it, we see that it is the 5th element as expected. Since `python_string` is a string, it can be indexed again, and we can access the 0th element.

Lastly, notice the bit of syntax in which two sets of brackets are placed. This is an efficient way to access something within a list or a tuple. 

## Part 3: Using Built-Ins
Create a variable called my_name and set it to be your name as a string:

In [ ]:
my_name = "<YOUR NAME HERE>"

We can combine strings together to make a full sentence:

In [ ]:
greeting = "hello world!"
print(greeting + " My name is " + my_name)

Notice that the sentence we printed isn't capitalized correctly. Luckily, Python  has a built-in function that can help us!  
To call the `capitalize()` function built into the Python string, put a dot after the string variable name:

In [ ]:
# Use the built-in method to capitalize the greeting
greeting = greeting.capitalize()
print(greeting + " My name is " + my_name)

##### Question: What do the following methods do to our string?

In [ ]:
print(greeting.upper(), greeting.lower(), greeting.replace('ll', 'y'))

##### Answer:
**YOUR ANSWER HERE**

Let's check out two more built-in functions to drive the point home: `split()` and `join()`.
- `split(char)`: returns a list of strings, fragments of the original string without the given `char`
- `join(split_string)`: inserts a string into a given list of strings, `split_string`

In [ ]:
split_greeting = greeting.split('o')
print("split string: ", split_greeting, "\ntype: ", type(split_greeting))

In [ ]:
joined_greeting = 'o'.join(split_greeting)
print("joined string: ", joined_greeting, "\ntype: ", type(joined_greeting))

If you ever forget what a built-in function does, hovering over the function with your cursor will let you view the Python documentation for that function.

## Part 4: Hex and Binary

Python does not have a built-in "bytes" data type like it does with integers and floats. You instead store binary data in a **bytes object**. We'll learn more about objects later, but just know that objects are created (aka *instantiated*) using a constructor.

Below, we encode the string 'hello' using ASCII encoding:  
(If you didn't learn ASCII encoding, this is a way of representing text using bytes)

In [1]:
byte_str = bytes('hello', 'ascii')

 The output below should be a **list** of the ASCII values for 'hello'.

In [2]:
list(byte_str)

[104, 101, 108, 108, 111]

We can use indexing to access individual bytes. We can use slicing to access a certain sequence of bytes.
   
(If you didn't know: ASCII 'h' = 0x68 = 104)

In [4]:
print(byte_str[0])

print(byte_str[1:4])

104
b'ell'


In the prior output, you might have noticed the `b''`. This is another way of declaring bytes:

In [5]:
print(b'my byte string')

print(bytes('my byte string', 'ascii'))

b'my byte string'
b'my byte string'


Finally, we often want to represent or work with bytes that are not just ASCII text. In order to declare raw bytes in our bytes object, we use the hex value of the byte and the escape sequence `\x`.  

For example, a byte string containing a single byte of 0xFF would look like this:

In [ ]:
my_byte = b'\xff'

print(my_byte)
print(my_byte.hex())

b'\xff'
ff


##### Question: Get the first 5 bytes from `gibberish_bytes`, and append to the beginning of the string.

In [ ]:
gibberish_bytes = b'\x47\x6fod\x20M\x37ilovepython'
my_string = "work!"

five_bytes = ???

new_string = ???.decode('ascii') + my_string
print(new_string)

Good work!


### Decoding and Encoding Bytes
Notice how we use the `.decode()` function built into the `bytes` object to convert from raw bytes to ASCII text. Python includes multiple built-ins to convert between bytes and other data types, such as text and integers.

To convert between text and bytes, use `encode()` and `decode()`.

In [ ]:
name_string = "Ben"

name_bytes = name_string.encode('ascii')
print(name_bytes)

new_name_string = name_bytes.decode('ascii')
print(new_name_string)

b'Ben'
Ben


You can format your bytes output as hex using `.hex()`

In [14]:
print(name_bytes.hex())

42656e


Use `to_bytes` and `from_bytes` to convert between bytes and integers:
- `num.to_bytes(n, endianness)` converts the `num` integer to a bytes object with a length of `n` bytes. 
- `int.from_bytes(bytes, endianness)` converts the  `bytes` object to an integer.

In [15]:
num = 1025
bytes = num.to_bytes(2, "big")
print(bytes)

b'\x04\x01'


In [16]:
num_recovered = int.from_bytes(bytes, "big")
print(num_recovered)

1025


`"big"` means big-endian (byte order is not reversed). For this week's lab, you might have to specify `"little"` instead if the byte data is in little-endian.

## That's All! Time for a Challenge...

If you're not a coward, you'll try your newly-acquired Python skills with this week's Python Lab.  

You have intercepted the internet traffic of a foreign adversary. We would like to know the exact details of their internet activity. You will receive the bytestring of one of their IPv4 packets. The packet is structured like so:

| Bit start .. bit end | Data | Type |
| :------------------- | :--: | ---: |
| 0..7  | Version  | Byte |
| 8..15 | TOS  | Byte  |
| 16..31| Total Length | Short (2 bytes) |
| 32..63 | Source Address | Int (4 bytes) |
| 64..96 | Destination Address | Int (4 bytes) |
| 97..352 | Data | (32 bytes) |

Each of these data fields, except for data, are in little-endian.

Download the template python script with `wget https://cyber.wildcats.cc/Week-3-Challenge.py`. Find the Version, TOS, Total Length, Source Address, and Destination Address. 

Once you do this, we will verify your answers. Download the verification script with `wget https://cyber.wildcats.cc/week-3-server`. Run the script by typing `./week-3-server` into your terminal, and input your answers.

Good luck, we're all counting on you.

